--- a/unix/xserver/hw/vnc/xvnc.c	2013-07-14 14:05:29.963390223 -0400
+++ b/unix/xserver/hw/vnc/xvnc.c	2013-07-14 14:04:12.840357191 -0400
@@ -250,7 +250,7 @@
 #if XORG < 111
 AbortDDX()
 #else
-AbortDDX(enum ExitCode error)
+SigAbortDDX(int signo, enum ExitCode error)
 #endif
 {
 #if XORG < 111
@@ -260,6 +260,14 @@
 #endif
 }
 
+#if XORG >= 111
+void
+AbortDDX(enum ExitCode error)
+{
+    SigAbortDDX(0, error);
+}
+#endif
+
 #ifdef __DARWIN__
 void
 DarwinHandleGUI(int argc, char *argv[])
