diff -bcr xc.org/programs/Xserver/Imakefile xc/programs/Xserver/Imakefile
*** xc.org/programs/Xserver/Imakefile	2004-12-15 20:22:53.000000000 +0100
--- xc/programs/Xserver/Imakefile	2005-02-14 14:23:21.000000000 +0100
***************
*** 128,134 ****
         LIBREGEX = RegexLibrary
  
  #if DoLoadableServer
!     LIBCWRAPPER = os/libcwrapper.o
  #endif
  
  #if BuildXprint
--- 128,134 ----
         LIBREGEX = RegexLibrary
  
  #if DoLoadableServer
!     LIBCWRAPPER = os/libcwrapper.o os/libos.a
  #endif
  
  #if BuildXprint
***************
*** 414,419 ****
--- 414,430 ----
  #define ServerToInstall Xsun
  #endif
  #endif /* XsunServer */
+ XCOMM
+ XCOMM X VNC server
+ XCOMM
+ FBSUBDIR = fb
+ XVNCDDXDIR = vnc/Xvnc
+ XVNCDIRS = $(STDDIRS) $(FBSUBDIR) $(XVNCDDXDIR) $(DEPDIRS)
+ XVNCLIBS = PreFbLibs vnc/Xvnc/LibraryTargetName(xvnc) FbPostFbLibs
+ XVNCOBJS = $(XVNCDDXDIR)/stubs.o $(XVNCDDXDIR)/miinitext.o
+ XVNCSYSLIBS = $(FONTLIBS) $(SYSLIBS)
+ ServerTarget(Xvnc,$(XVNCDIRS),$(XVNCOBJS), \
+ 	$(XVNCLIBS) $(LOADABLEEXTS) $(LIBCWRAPPER),$(XVNCSYSLIBS))
  
  
  #if defined(Xsun24Server) && Xsun24Server
diff -bcr xc.org/programs/Xserver/mi/miinitext.c xc/programs/Xserver/mi/miinitext.c
*** xc.org/programs/Xserver/mi/miinitext.c	2004-12-13 06:48:41.000000000 +0100
--- xc/programs/Xserver/mi/miinitext.c	2005-02-11 16:47:34.000000000 +0100
***************
*** 286,291 ****
--- 286,294 ----
  #ifdef MITMISC
  extern void MITMiscExtensionInit(INITARGS);
  #endif
+ #ifdef VNCEXT
+ extern void vncExtensionInit(INITARGS);
+ #endif
  #ifdef XIDLE
  extern void XIdleExtensionInit(INITARGS);
  #endif
***************
*** 587,592 ****
--- 590,598 ----
  #ifdef MITMISC
      if (!noMITMiscExtension) MITMiscExtensionInit();
  #endif
+ #ifdef VNCEXT
+     vncExtensionInit();
+ #endif
  #ifdef XIDLE
      if (!noXIdleExtension) XIdleExtensionInit();
  #endif
