--- freetype-2.3.11/src/base/ftoutln.c	2009-03-14 14:45:26.000000000 +0100
+++ freetype-2.3.11/src/base/ftoutln.c	2012-04-03 11:03:35.000000000 +0200
@@ -990,7 +990,8 @@
 
     int             i;
     FT_Pos          ray_y[3];
-    FT_Orientation  result[3];
+    FT_Orientation  result[3] =
+      { FT_ORIENTATION_NONE, FT_ORIENTATION_NONE, FT_ORIENTATION_NONE };
 
 
     if ( !outline || outline->n_points <= 0 )
