--- freetype-2.3.11/ft2demos-2.3.11/src/ftmulti.c	2010-07-22 19:11:50.000000000 +0200
+++ freetype-2.3.11/ft2demos-2.3.11/src/ftmulti.c	2010-07-22 19:12:41.000000000 +0200
@@ -813,13 +813,13 @@
 
           for ( n = 0; n < (int)multimaster->num_axis; n++ )
           {
-            char  temp[32];
+            char  temp[100];
 
 
-            sprintf( temp, "  %s:%g",
+            sprintf( temp, "  %.50s:%g",
                            multimaster->axis[n].name,
-                           design_pos[n]/65536. );
-            strcat( Header, temp );
+                           design_pos[n] / 65536.0 );
+            strncat( Header, temp, sizeof( Header ) - strlen( Header ) - 1 );
           }
         }
         grWriteCellString( &bit, 0, 16, Header, fore_color );
