diff -up freetype-2.3.11/ft2demos-2.3.11/Makefile.more-demos freetype-2.3.11/ft2demos-2.3.11/Makefile
--- freetype-2.3.11/ft2demos-2.3.11/Makefile.more-demos	2009-10-22 16:02:26.000000000 -0400
+++ freetype-2.3.11/ft2demos-2.3.11/Makefile	2009-10-22 16:02:32.000000000 -0400
@@ -288,10 +288,10 @@ else
   # Note that ttdebug only works if the FreeType's `truetype' driver has
   # been compiled with TT_CONFIG_OPTION_BYTECODE_INTERPRETER defined.
   #
-  # EXES += ftchkwd
-  # EXES += ftmemchk
-  # EXES += ftpatchk
-  # EXES += fttimer
+  EXES += ftchkwd
+  EXES += ftmemchk
+  EXES += ftpatchk
+  EXES += fttimer
   # EXES += testname
   # EXES += ttdebug
 
