--- freetype-2.2.1/builds/unix/freetype-config.in.multilib	2006-07-27 18:50:40.000000000 -0400
+++ freetype-2.2.1/builds/unix/freetype-config.in	2006-07-27 18:58:13.000000000 -0400
@@ -9,11 +9,11 @@
 # indicate that you have read the license and understand and accept it
 # fully.
 
-prefix=@prefix@
-exec_prefix=@exec_prefix@
+prefix=`pkg-config --variable prefix freetype2`
+exec_prefix=`pkg-config --variable exec_prefix freetype2`
 exec_prefix_set=no
-includedir=@includedir@
-libdir=@libdir@
+includedir=`pkg-config --variable includedir freetype2`
+libdir=`pkg-config --variable libdir freetype2`
 enable_shared=@build_libtool_libs@
 wl=@wl@
 hardcode_libdir_flag_spec='@hardcode_libdir_flag_spec@'
