diff -up tigervnc-1.0.90-20100721svn4113/unix/xserver/hw/vnc/Makefile.am.ldnow tigervnc-1.0.90-20100721svn4113/unix/xserver/hw/vnc/Makefile.am
--- tigervnc-1.0.90-20100721svn4113/unix/xserver/hw/vnc/Makefile.am.ldnow	2010-05-18 15:48:02.000000000 +0200
+++ tigervnc-1.0.90-20100721svn4113/unix/xserver/hw/vnc/Makefile.am	2010-07-21 17:05:45.242942531 +0200
@@ -50,7 +50,7 @@ libvnc_la_CPPFLAGS = $(XVNC_CPPFLAGS) -I
 	-I$(top_srcdir)/hw/xfree86/os-support/bus \
 	-I$(includedir)/pixman-1 -I$(includedir)
 
-libvnc_la_LDFLAGS = -module -avoid-version
+libvnc_la_LDFLAGS = -module -avoid-version -Wl,-z,now
 
 libvnc_la_LIBADD = libvnccommon.la $(COMMON_LIBS)
 
