diff --git a/m4/shave.m4 b/m4/shave.m4
index 01cb5c7..0dfde6c 100644
--- a/m4/shave.m4
+++ b/m4/shave.m4
@@ -32,8 +32,8 @@ AC_DEFUN([SHAVE_INIT],
   if test x"$enable_shave" = xyes; then
     dnl where can we find the shave scripts?
     m4_if([$1],,
-      [shavedir="$ac_pwd"],
-      [shavedir="$ac_pwd/$1"])
+      [shavedir=`pwd`],
+      [shavedir=`pwd`/$1])
     AC_SUBST(shavedir)
 
     dnl make is now quiet
