--- a/cmake/StaticBuild.cmake	2014-10-29 20:23:24.000000000 -0400
+++ b/cmake/StaticBuild.cmake	2014-10-29 20:24:05.000000000 -0400
@@ -68,7 +68,7 @@
   endif()
 
   if(FLTK_FOUND)
-    set(FLTK_LIBRARIES "-Wl,-Bstatic -lfltk_images -lpng -ljpeg -lfltk -Wl,-Bdynamic")
+    set(FLTK_LIBRARIES "-Wl,-Bstatic -L${FLTK_LIBRARY_DIR} -lfltk_images -lpng -ljpeg -lfltk -Wl,-Bdynamic")
 
     if(WIN32)
       set(FLTK_LIBRARIES "${FLTK_LIBRARIES} -lcomctl32")
--- a/vncviewer/CMakeLists.txt	2014-10-29 20:39:58.000000000 -0400
+++ b/vncviewer/CMakeLists.txt	2014-10-29 21:01:31.000000000 -0400
@@ -46,7 +46,7 @@
   add_executable(vncviewer ${VNCVIEWER_SOURCES})
 endif()
 
-target_link_libraries(vncviewer rfb network rdr os Xregion ${FLTK_LIBRARIES} ${GETTEXT_LIBRARIES})
+target_link_libraries(vncviewer ${FLTK_LIBRARIES} rfb network rdr os Xregion -lxcb ${X11_Xdmcp_LIB} ${X11_Xau_LIB} ${GETTEXT_LIBRARIES})
 
 if(APPLE)
   target_link_libraries(vncviewer "-framework Cocoa" "-framework Carbon")
