Index: fb/fbblt.c
================================================================================
--- fb/fbblt.c
+++ fb/fbblt.c
@@ -76,6 +76,11 @@
     }
 #endif
 
+/* 
+ * Disable optimization (introduced by ajax) due to a general vesa 
+ * driver crash later in memcpy 
+*/
+#if 0
     if (alu == GXcopy && pm == FB_ALLONES && !reverse &&
             !(srcX & 7) && !(dstX & 7) && !(width & 7)) {
         int i;
@@ -97,6 +102,7 @@
 
         return;
     }
+#endif
 
     FbInitializeMergeRop(alu, pm);
     destInvarient = FbDestInvarientMergeRop();
