Index: hw/xfree86/common/xf86Helper.c
================================================================================
--- hw/xfree86/common/xf86Helper.c
+++ hw/xfree86/common/xf86Helper.c
@@ -2183,6 +2183,13 @@
 			&errmaj, &errmin);
     if (!ret)
 	LoaderErrorMsg(pScrn->name, name, errmaj, errmin);
+
+    /* "drm" module has been merged into "dri" module, but some drivers
+        are still trying to load it explictely and fail if it doesn't exist :-( */
+    if (!ret && !strcmp(name,"drm")) {
+        ret = (pointer) -1;
+    }
+
     return ret;
 }
 
