--- configure.ac
+++ configure.ac
@@ -505,11 +505,15 @@
 XORG_FONTSUBDIR(FONTTYPE1DIR, fonttype1dir, Type1)
 XORG_FONTSUBDIR(FONT75DPIDIR, font75dpidir, 75dpi)
 XORG_FONTSUBDIR(FONT100DPIDIR, font100dpidir, 100dpi)
+XORG_FONTSUBDIR(FONTURWDIR, fonturwdir, URW)
+XORG_FONTSUBDIR(FONTCYRILLICDIR, fontcyrillicdir, cyrillic)
+XORG_FONTSUBDIR(FONTSGIDIR, fontsgidir, misc/sgi)
+XORG_FONTSUBDIR(FONTTRUETYPEDIR, fonttruetypedir, truetype)
 
 dnl Uses --default-font-path if set, otherwise checks for /etc/X11/fontpath.d,
 dnl otherwise uses standard subdirectories of FONTROOTDIR. When cross
 dnl compiling, assume default font path uses standard FONTROOTDIR directories.
-DEFAULT_FONT_PATH="${FONTMISCDIR}:unscaled,${FONTTTFDIR}/,${FONTOTFDIR}/,${FONTTYPE1DIR}/,${FONT100DPIDIR}/,${FONT75DPIDIR}/"
+DEFAULT_FONT_PATH="${FONTMISCDIR}:unscaled,${FONTTTFDIR}/,${FONTOTFDIR}/,${FONTTYPE1DIR}/,${FONT100DPIDIR}:unscaled,${FONT75DPIDIR}:unscaled,${FONTURWDIR}/,${FONTCYRILLICDIR}:unscaled,${FONTSGIDIR}:unscaled,${FONTTRUETYPEDIR}/"
 if test "$cross_compiling" != yes; then
 	AC_CHECK_FILE([${sysconfdir}/X11/fontpath.d],
 		[DEFAULT_FONT_PATH='catalogue:${sysconfdir}/X11/fontpath.d'],
