Index: hw/xfree86/common/xf86Globals.c
================================================================================
--- hw/xfree86/common/xf86Globals.c
+++ hw/xfree86/common/xf86Globals.c
@@ -144,7 +144,7 @@
 };
 const char *xf86ConfigFile = NULL;
 const char *xf86ConfigDir = NULL;
-const char *xf86ModulePath = DEFAULT_MODULE_PATH;
+const char *xf86ModulePath = DEFAULT_MODULE_PATH "/updates," DEFAULT_MODULE_PATH;
 MessageType xf86ModPathFrom = X_DEFAULT;
 const char *xf86LogFile = DEFAULT_LOGPREFIX;
 MessageType xf86LogFileFrom = X_DEFAULT;
