Index: xorg-server-1.6.3.901/hw/xfree86/common/xf86RandR.c
================================================================================
--- xorg-server-1.7.99/hw/xfree86/common/xf86RandR.c
+++ xorg-server-1.7.99/hw/xfree86/common/xf86RandR.c
@@ -247,6 +247,9 @@
     Bool		    useVirtual = FALSE;
     Rotation		    oldRotation = randrp->rotation;
 
+    if (!scrp->vtSema)
+	return FALSE;
+
     miPointerGetPosition(inputInfo.pointer, &px, &py);
     for (mode = scrp->modes; ; mode = mode->next)
     {
