--- xorg-server-1.9.3/dix/dixfonts.c.orig	2010-10-31 03:47:09.000000000 +0100
+++ xorg-server-1.9.3/dix/dixfonts.c	2013-10-28 16:12:38.696658572 +0100
@@ -1500,6 +1500,7 @@
 	    GC *pGC;
 	    unsigned char *data;
 	    ITclosurePtr new_closure;
+	    ITclosurePtr old_closure;
 
 	    /* We're putting the client to sleep.  We need to
 	       save some state.  Similar problem to that handled
@@ -1512,6 +1513,7 @@
 		err = BadAlloc;
 		goto bail;
 	    }
+	    old_closure = c;
 	    *new_closure = *c;
 	    c = new_closure;
 
@@ -1519,6 +1521,7 @@
 	    if (!data)
 	    {
 		free(c);
+		c = old_closure;
 		err = BadAlloc;
 		goto bail;
 	    }
@@ -1530,6 +1533,7 @@
 	    {
 		free(c->data);
 		free(c);
+		c = old_closure;
 		err = BadAlloc;
 		goto bail;
 	    }
@@ -1543,6 +1547,7 @@
 		FreeScratchGC(pGC);
 		free(c->data);
 		free(c);
+		c = old_closure;
 		err = BadAlloc;
 		goto bail;
 	    }
