Index: hw/kdrive/Makefile.am
================================================================================
--- hw/kdrive/Makefile.am
+++ hw/kdrive/Makefile.am
@@ -1,5 +1,5 @@
 if BUILD_KDRIVEFBDEVLIB
-FBDEV_SUBDIRS = fbdev
+FBDEV_SUBDIRS =
 endif
 
 if XFAKESERVER
@@ -16,8 +16,7 @@
 
 SERVER_SUBDIRS = 		\
 	$(FBDEV_SUBDIRS)	\
-	$(XEPHYR_SUBDIRS)       \
-	$(XFAKE_SUBDIRS)
+	$(XEPHYR_SUBDIRS)
 
 SUBDIRS =			\
 	src			\
