--- man/Makefile.am.orig	2010-09-14 01:25:46.000000000 +0200
+++ man/Makefile.am	2010-11-12 13:28:31.000000000 +0100
@@ -23,12 +23,8 @@
 miscmandir = $(MISC_MAN_DIR)
 
 miscman_PRE =			\
-	Consortium.man		\
 	Xsecurity.man		\
-	Standards.man		\
-	X.man			\
-	XOrgFoundation.man	\
-	XProjectTeam.man
+	X.man
 
 miscman_DATA = $(miscman_PRE:man=@MISC_MAN_SUFFIX@)
 
--- Makefile.am.orig	2010-11-12 13:45:19.000000000 +0100
+++ Makefile.am	2010-11-12 13:45:32.000000000 +0100
@@ -20,9 +20,7 @@
 #
 #  Process this file with autoconf to create configure.
 
-SUBDIRS = general man specs
-
-dist_doc_DATA = MAINTAINERS
+SUBDIRS = man
 
 MAINTAINERCLEANFILES = ChangeLog INSTALL
 EXTRA_DIST = \
