Index: xorg-server-1.6.3.901/hw/xfree86/common/xf86AutoConfig.c
================================================================================
--- xorg-server-1.7.99/hw/xfree86/common/xf86AutoConfig.c
+++ xorg-server-1.7.99/hw/xfree86/common/xf86AutoConfig.c
@@ -535,7 +535,8 @@
 #if !defined(__linux__) && defined(__sparc__)
 	matches[i++] = xnfstrdup("wsfb");
 #else
-	matches[i++] = xnfstrdup("vesa");
+	if (geteuid() == 0)
+	    matches[i++] = xnfstrdup("vesa");
 #endif
     }
 }
