--- fltk-1.3.2.org/src/Fl_cocoa.mm      2013-01-16 11:32:11.788478228 +0100
+++ fltk-1.3.2/src/Fl_cocoa.mm  2013-01-16 11:32:55.824101285 +0100
@@ -3727,7 +3727,7 @@ CGImageRef Fl_X::CGImage_from_window_rec
   CGImageRef img;
   if (fl_mac_os_version >= 100500) {
     NSBitmapImageRep *bitmap = rect_to_NSBitmapImageRep(win, x, y, w, h);
-    img = [bitmap CGImage]; // requires Mac OS 10.5
+    img = (CGImageRef)[bitmap CGImage]; // requires Mac OS 10.5
     CGImageRetain(img);
     [bitmap release];
     }
